src/Framework/AspNetCoreAnalyzers/src/Analyzers/Authorization/AddAuthorizationBuilderAnalyzer.cs

```diff
@@ -0,0 +1,213 @@
+        }
+
+        var addAuthorizationMethod = policyServiceCollectionExtensions.GetMembers()
+            .OfType<IMethodSymbol>()
+            .FirstOrDefault(member => member is { Name: "AddAuthorization", Parameters.Length: 2 });
+
+        if (addAuthorizationMethod is null)
+        {
+            return;
+        }
+
+        context.RegisterOperationAction(context =>
+        {
+            var invocation = (IInvocationOperation)context.Operation;
+
+            if (SymbolEqualityComparer.Default.Equals(invocation.TargetMethod, addAuthorizationMethod)
+                && SymbolEqualityComparer.Default.Equals(invocation.TargetMethod.ContainingType, policyServiceCollectionExtensions)
+                && IsLastCallInChain(invocation)
+                && IsCompatibleWithAuthorizationBuilder(invocation, authorizationOptionsTypes))
+            {
+                AddDiagnosticInformation(context, invocation.Syntax.GetLocation());
+            }
+
+        }, OperationKind.Invocation);
+    }
+
+    private static bool IsCompatibleWithAuthorizationBuilder(IInvocationOperation invocation, AuthorizationOptionsTypes authorizationOptionsTypes)
+    {
+        if (invocation is { Arguments: { Length: 2 } invocationArguments }
+            && invocationArguments[1] is IArgumentOperation { ChildOperations: { Count: 1 } argumentOperations }
+            && argumentOperations.First() is IDelegateCreationOperation { ChildOperations: { Count: 1 } delegateCreationOperations }
+            && delegateCreationOperations.First() is IAnonymousFunctionOperation { ChildOperations: { Count: 1 } anonymousFunctionOperations }
+            && anonymousFunctionOperations.First() is IBlockOperation blockOperation)
+        {
+            // Ensure that the child operations of the configuration action passed to AddAuthorization are all related to AuthorizationOptions.
+            var allOperationsInvolveAuthorizationOptions = blockOperation.ChildOperations
+                .Where(operation => operation is not IReturnOperation { IsImplicit: true })
+                .All(operation => DoesOperationInvolveAuthorizationOptions(operation, authorizationOptionsTypes));
+
+            return allOperationsInvolveAuthorizationOptions
+                // Ensure that the configuration action passed to AddAuthorization does not use any AuthorizationOptions-specific APIs.
+                && IsConfigureActionCompatibleWithAuthorizationBuilder(blockOperation, authorizationOptionsTypes);
+        }
+
+        return false;
+    }
+
+    private static bool DoesOperationInvolveAuthorizationOptions(IOperation operation, AuthorizationOptionsTypes authorizationOptionsTypes)
+    {
+        if (operation is IExpressionStatementOperation { Operation: { } expressionStatementOperation })
+        {
+            if (expressionStatementOperation is ISimpleAssignmentOperation { Target: IPropertyReferenceOperation { Property.ContainingType: { } propertyReferenceContainingType } }
```

Comment:
Do we need add tests that what these code segments are checking (assuming I understood the code correctly) like:

```csharp
builder.Services.AddAuthorization(options => ConfigureOptions(options));
```

Reply:
I've added some new tests and updated a few of the existing ones to ensure that all of the branches in these private methods are covered by tests. I also removed the `MethodReferenceOperation`-related branch as this was unreachable.